### PR TITLE
[data] sanitize streaming executor resource management

### PR DESCRIPTION
**Pull request by raulchen**
## Why are these changes needed?

Currently we have multiple code places managing resource usages for different components of the streaming executor. 

This PR sanitize the code by consolidating them into one single class `ResourceManager`.  No code logic changes in this PR.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/execution/streaming_executor_state.py [hunk 1]
```diff
@@ -684,24 +627,23 @@ def _execution_allowed(
 
     Args:
         op: The operator to check.
-        global_usage: Resource usage across the entire topology.
```
Comment (bveeramani) [inline python/ray/data/_internal/execution/streaming_executor_state.py:687]: Add parameter description for `resource_manager`?

**Review by bveeramani — APPROVED** [review]

File: python/ray/data/_internal/execution/resource_manager.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            gpu=gpu,
+            object_store_memory=object_store_memory,
+        )
+        return self._global_limits
+
+    def get_op_usage(self, op: PhysicalOperator) -> ExecutionResources:
+        """Return the resource usage of the given operator at the current time."""
+        return self._op_usages[op]
+
+    def get_downstream_fraction(self, op: PhysicalOperator) -> float:
```
Comment (c21) [inline python/ray/data/_internal/execution/resource_manager.py:117]: `downstream_fraction` is a bit vague term. Can we call it something like `get_downstream_usage_fraction`?

File: python/ray/data/_internal/execution/resource_manager.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    ExecutionResources,
+)
+from ray.data._internal.execution.interfaces.physical_operator import PhysicalOperator
+from ray.data._internal.execution.operators.input_data_buffer import InputDataBuffer
+
+if TYPE_CHECKING:
+    from ray.data._internal.execution.streaming_executor_state import Topology
+
+
+class ResourceManager:
```
Comment (c21) [inline python/ray/data/_internal/execution/resource_manager.py:16]: It's mostly a `ResourceMonitor` / `ResourceTracker`. `ResourceManager` makes me think if they are responsible for scheduling resource (launch/kill tasks).

File: python/ray/data/_internal/execution/resource_manager.py (hunk 2 above)
Comment (raulchen) [reply python/ray/data/_internal/execution/resource_manager.py:117]: this fraction is actually not based on usage, it's based on number of operators, e.g., the last op is  `1/N`, second last is `2/N`. 
I agree the term is a bit vague, but we used the same term previously as well.

File: python/ray/data/_internal/execution/resource_manager.py (hunk 3 above)
Comment (raulchen) [reply python/ray/data/_internal/execution/resource_manager.py:16]: I actually plan to add the per-op resource allocation to this class in the next PR. At that time, monitor or tracker will be inaccurate. Thus I prefer keeping it for now. If we consider it as "managing the resource usages", it still kind of makes sense

Comment (c21) [reply python/ray/data/_internal/execution/resource_manager.py:16]: SG

File: python/ray/data/_internal/execution/resource_manager.py (hunk 2 above)
Comment (c21) [reply python/ray/data/_internal/execution/resource_manager.py:117]: Got it, thanks.

**Review by c21 — APPROVED** [review]

**Pull request merged**